test/e2e/apps/job.go

```diff
@@ -140,76 +141,29 @@ var _ = SIGDescribe("Job", func() {
-		the non-0-indexed pods to determine if the marker file is already
-		created by the 0th indexed pod - the non-0-indexed pods fail and restart
-		until the marker file is created (their potential failures are ignored
-		based on the exit code). Once the marker file is created the 0th indexed
-		pod is evicted (DisruptionTarget condition is added in the process),
-		after restart it runs to successful completion.
-		Steps:
-		1. Select a node to run all Job's pods to ensure the host marker file is accessible by all pods
-		2. Create the indexed job with pod failure policy which ignores failed pods with 137 exit code
-		3. Await for all non-0-indexed pods to succeed to ensure the marker file is created by the 0-indexed pod
-		4. Make sure the 0-indexed pod is running
-		5. Evict the 0-indexed pod, the failure is ignored as it matches the pod failure policy
-		6. Await for the job to successfully complete
+		Testname: Ensure pod failure policy allows to ignore failure matching on the exit code
+		Description: This test is using an indexed job. The pod corresponding to each index
+		creates a marker file on the host and runs 'forever' until evicted. Once
+		the marker file is created the pod succeeds seeing it on restart. Thus,
+		we trigger one failure per index due to eviction, so the Job would be
+		marked as failed, if not for the ignore rule matching on exit codes.
 	*/
-	ginkgo.It("should allow to use a pod failure policy to ignore failure for an evicted pod; matching on the exit code", func(ctx context.Context) {
-		// We set the backoffLimit to 0 so that any pod failure would trigger
-		// job failure if not for the pod failure policy to ignore the failed
-		// pods from counting them towards the backoffLimit.
-		parallelism := int32(2)
-		completions := int32(4)
-		backoffLimit := int32(0)
+	ginkgo.It("should allow to use a pod failure policy to ignore failure matching on exit code", func(ctx context.Context) {
+		// We set the backoffLimit = numPods-1  so that we can tolerate random
+		// failures (like OutOfPods from kubelet). Yet, the Job would fail if the
+		// pod failures were not be ignored.
+		numPods := 3
+		parallelism := int32(numPods)
+		completions := int32(numPods)
+		backoffLimit := int32(numPods) - 1
 
 		ginkgo.By("Looking for a node to schedule job pods")
 		node, err := e2enode.GetRandomReadySchedulableNode(ctx, f.ClientSet)
 		framework.ExpectNoError(err)
 
 		ginkgo.By("Creating a job")
-		job := e2ejob.NewTestJobOnNode("notTerminateOnce", "evicted-pod-ignore-on-exit-code", v1.RestartPolicyNever, parallelism, completions, nil, backoffLimit, node.Name)
+		job := e2ejob.NewTestJobOnNode("notTerminateOncePerIndex", "evicted-pod-ignore-on-exit-code", v1.RestartPolicyNever, parallelism, completions, nil, backoffLimit, node.Name)
 		job.Spec.CompletionMode = ptr.To(batchv1.IndexedCompletion)
 		job.Spec.PodFailurePolicy = &batchv1.PodFailurePolicy{
 			Rules: []batchv1.PodFailurePolicyRule{
```

Comment:
the comment bellow needs an update

Reply:
also an observation: exit code 1 will not be observed. Is this intentional just to have the more values set? Might be usefulo to comment on that.

Reply:
I added exit code 1 as it was a potential scenario in the past: https://github.com/kubernetes/kubernetes/blob/20dd88872deb4516a6f8cbc32a59619951680c53/test/e2e/framework/job/fixtures.go#L157. However, we can simplify now. I will also delete `notTerminateOnce` as no longer used. It was more complex, so I'm happy we are able to get rid of it.

Reply:
okay, sounds good to me